workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+```haskell
+identity :: Int -> Int
+identity x = x
+
+hello :: String -> String
+hello name = "Hello, " ++ name
+
+addInt :: Int -> Int -> Int
+addInt x y = x + y
+```
+
+Add these ^ to a file and try them out!
+
+## A First Function
+
+* A function that converts (some) integers to words
+* Pattern matching
+
+```haskell
+intToWord :: Int -> String
+intToWord 1 = "one"
+intToWord 2 = "two"
+intToWord 3 = "three"
+intToWord _ = "dunno"
+```
+
+* Try and write `wordToInt :: String -> Int`
+
+## Polymorphic Types
+
+* Some functions have the same behaviour for values of different types
+  * E.g. the `identity` function
+* Can generalise functions by using *type variables*
+
+```haskell
+identity :: a -> a
+identity x = x
+```
+
+* Can also *constrain* the type variables
+
+```haskell
+add :: Num a => a -> a -> a
+add x y = x + y
+```
+
+* These constraints can be defined using *type classes*
+
+## Composition
+
+* Using the output of one function as the input of another function: *"chaining functions together"*
+* Use the dot operator for composition
+
+```haskell
+f :: Int -> String
+f x = intToWord x
+
+g :: String -> Int
+g x = wordToInt x
+
+h :: Int -> Int
+h = g . f
+
+(.) :: (b -> c) -> (a -> b) -> a -> c
+(.) f g x = g (f x)
+```
+
+
+# The Data Model
+
+## Goal
+
+* Create a data structure that accurately models the data
+* Introduce Maybe
+* Introduce Lists
+
+### Approach
+
+* Define a model for a single row of data (an average candidate)
+* Apply this model to all the rows
+
+![](images/data-model.pdf)
+
+## A Single Candidate
+
+![](images/small-data.png)
+
+* A candidate has 4 attributes:
+  - constituency
+  - party
+  - sex
+  - average age
+
+* Need a way of grouping these together
+
+## Data Types
+
+A simple data type:
+
+```haskell
+data Person = MakePerson
+  { name :: String
+  , age  :: Int
+  } deriving Show
+```
+
+* `Person` is the type being defined
+* `MakePerson` is the constructor function
+* `name, age` are fields with associated types
+* `deriving Show` allows the use of the `show` function
+  * `show` converts structure to a string
+
+## Working with Data Types
+
+* Creating a new data type:
+
+```haskell
+MakePerson :: String -> Int -> Person
+```
+* Accessing the data type:
+
+```haskell
+name :: Person -> String
+age :: Person -> Int
+```
+![](images/simple-datatype.png)
+
+## The Maybe Type
+
+* What happens if we cannot parse a candidate?
+* Need some way to represent an invalid row:
+
+```haskell
+data Maybe a = Nothing | Just a
+```
+
+* `Maybe` is the type being defined
+* `a` is a *type variable* (could be any type)
+* `Nothing, Just` are *both* constructors
+  * `Nothing :: Maybe a`
+  * `Just :: a -> Maybe a`
+
+## The List Type
+
+* How do we represent a list of candidates?
+* Use singly-linked lists:
+  * `[]` is the empty list
+  * `:` lets you append values to the front
+  * `[a,b,c]` == `(a : b : c : [])`
+
+![](images/list.png){ height=40% }
+
+## The Data Model
+
+* Define our basic candidate type
```

Comment:
Add a presenter note that we should pause here and let the participants to try out their implementation, so that we don't accidentally go to the next page

Reply:
Done